File: src/ios/FirebaseAuthenticationPlugin.m
#import "FirebaseAuthenticationPlugin.h"

@implementation FirebaseAuthenticationPlugin

- (void)pluginInitialize {
    NSLog(@"Starting Firebase Authentication plugin");

    if(![FIRApp defaultApp]) {
        [FIRApp configure];
    }
}

- (void)setAuthStateChanged:(CDVInvokedUrlCommand*)command {
    BOOL disable = [[command.arguments objectAtIndex:0] boolValue];
    if (self.authChangedHandler) {
        [[FIRAuth auth] removeAuthStateDidChangeListener:self.authChangedHandler];
        self.authChangedHandler = nil;
    }
    if (!disable) {
        self.authChangedHandler = [[FIRAuth auth] addAuthStateDidChangeListener:^(FIRAuth* auth, FIRUser* user) {
           CDVPluginResult *pluginResult = [self getProfileResult:user];
           [pluginResult setKeepCallbackAsBool:YES];
           [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
       }];
    }
}

- (void)getCurrentUser:(CDVInvokedUrlCommand *)command {
    CDVPluginResult *pluginResult = [self getProfileResult:[FIRAuth auth].currentUser];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)getIdToken:(CDVInvokedUrlCommand *)command {
    BOOL forceRefresh = [[command.arguments objectAtIndex:0] boolValue];
    FIRUser *user = [FIRAuth auth].currentUser;

    if (user) {
        [user getIDTokenForcingRefresh:forceRefresh completion:^(NSString *token, NSError *error) {
            CDVPluginResult *pluginResult;
            if (error) {
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.localizedDescription];
            } else {
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:token];
            }

            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }];
    } else {
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"User must be signed in"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }
}

- (void)createUserWithEmailAndPassword:(CDVInvokedUrlCommand *)command {
    NSString* email = [command.arguments objectAtIndex:0];
    NSString* password = [command.arguments objectAtIndex:1];

    [[FIRAuth auth] createUserWithEmail:email
                               password:password
                             completion:^(FIRAuthDataResult *result, NSError *error) {
        [self respondWith:error callbackId:command.callbackId];
    }];
}

- (void)sendEmailVerification:(CDVInvokedUrlCommand *)command {
    FIRUser *currentUser = [FIRAuth auth].currentUser;

    if (currentUser) {
        [currentUser sendEmailVerificationWithCompletion:^(NSError *_Nullable error) {
            [self respondWith:error callbackId:command.callbackId];
        }];
    } else {
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"User must be signed in"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }
}

- (void)sendPasswordResetEmail:(CDVInvokedUrlCommand *)command {
    NSString* email = [command.arguments objectAtIndex:0];

    [[FIRAuth auth] sendPasswordResetWithEmail:email completion:^(NSError *_Nullable error) {
        [self respondWith:error callbackId:command.callbackId];
    }];
}

- (void)signInWithEmailAndPassword:(CDVInvokedUrlCommand *)command {
    NSString* email = [command.arguments objectAtIndex:0];
    NSString* password = [command.arguments objectAtIndex:1];

    [[FIRAuth auth] signInWithEmail:email
                           password:password
                         completion:^(FIRAuthDataResult *result, NSError *error) {
        [self respondWith:error callbackId:command.callbackId];
    }];
}

- (void)signInAnonymously:(CDVInvokedUrlCommand *)command {
    [[FIRAuth auth] signInAnonymouslyWithCompletion:^(FIRAuthDataResult *result, NSError *error) {
        [self respondWith:error callbackId:command.callbackId];
    }];
}

- (void)signInWithGoogle:(CDVInvokedUrlCommand *)command {
    NSString* idToken = [command.arguments objectAtIndex:0];
    NSString* accessToken = [command.arguments objectAtIndex:1];

    FIRAuthCredential *credential = [FIRGoogleAuthProvider credentialWithIDToken:idToken
                                                                     accessToken:accessToken];
    [[FIRAuth auth] signInWithCredential:credential
                              completion:^(FIRAuthDataResult *result, NSError *error) {
        [self respondWith:error callbackId:command.callbackId];
    }];
}

- (void)signInWithFacebook:(CDVInvokedUrlCommand *)command {
    NSString* accessToken = [command.arguments objectAtIndex:0];

    FIRAuthCredential *credential = [FIRFacebookAuthProvider credentialWithAccessToken:accessToken];
    [[FIRAuth auth] signInWithCredential:credential
                              completion:^(FIRAuthDataResult *result, NSError *error) {
        [self respondWith:error callbackId:command.callbackId];
    }];
}

- (void)signInWithTwitter:(CDVInvokedUrlCommand *)command {
    NSString* token = [command.arguments objectAtIndex:0];
    NSString* secret = [command.arguments objectAtIndex:1];

    FIRAuthCredential *credential = [FIRTwitterAuthProvider credentialWithToken:token
                                                                         secret:secret];
    [[FIRAuth auth] signInWithCredential:credential
                              completion:^(FIRAuthDataResult *result, NSError *error) {
        [self respondWith:error callbackId:command.callbackId];
    }];
}

- (void)signInWithApple:(CDVInvokedUrlCommand *)command {
    NSString* idToken = [command.arguments objectAtIndex:0];
    NSString* rawNonce = [command.arguments objectAtIndex:1];

    FIRAuthCredential *credential = [FIROAuthProvider credentialWithProviderID:@"apple.com"
                                                                       IDToken:idToken
                                                                      rawNonce:rawNonce];
    [[FIRAuth auth] signInWithCredential:credential
                              completion:^(FIRAuthDataResult *result, NSError *error) {
        [self respondWith:error callbackId:command.callbackId];
    }];
}

- (void)signInWithVerificationId:(CDVInvokedUrlCommand*)command {
    NSString* verificationId = [command.arguments objectAtIndex:0];
    NSString* smsCode = [command.arguments objectAtIndex:1];

    FIRAuthCredential *credential = [[FIRPhoneAuthProvider provider]
            credentialWithVerificationID:verificationId
                        verificationCode:smsCode];

    [[FIRAuth auth] signInWithCredential:credential
                              completion:^(FIRAuthDataResult *result, NSError *error) {
        [self respondWith:error callbackId:command.callbackId];
    }];
}

- (void)verifyPhoneNumber:(CDVInvokedUrlCommand*)command {
    NSString* phoneNumber = [command.arguments objectAtIndex:0];

    [[FIRPhoneAuthProvider provider] verifyPhoneNumber:phoneNumber
                                            UIDelegate:nil
                                            completion:^(NSString* verificationId, NSError* error) {
        CDVPluginResult *pluginResult;
        if (error) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.localizedDescription];
        } else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:verificationId];
        }
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void)signInWithCustomToken:(CDVInvokedUrlCommand *)command {
    NSString* idToken = [command.arguments objectAtIndex:0];

    [[FIRAuth auth] signInWithCustomToken:idToken
                         completion:^(FIRAuthDataResult *result, NSError *error) {
        [self respondWith:error callbackId:command.callbackId];
    }];
}

- (void)signOut:(CDVInvokedUrlCommand*)command {
    NSError *signOutError;
    CDVPluginResult *pluginResult;

    if ([[FIRAuth auth] signOut:&signOutError]) {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    } else {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:signOutError.localizedDescription];
    }
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)setLanguageCode:(CDVInvokedUrlCommand*)command {
    NSString* languageCode = [command.arguments objectAtIndex:0];
    if (languageCode) {
        [FIRAuth auth].languageCode = languageCode;
    } else {
        [[FIRAuth auth] useAppLanguage];
    }

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)updateProfile:(CDVInvokedUrlCommand*)command {
    NSDictionary* options = [command argumentAtIndex:0];
    FIRUserProfileChangeRequest* changeRequest = [[FIRAuth auth].currentUser profileChangeRequest];
    if (options[@"displayName"]) {
        changeRequest.displayName = options[@"displayName"];
    }
    if (options[@"photoURL"]) {
        changeRequest.photoURL = options[@"photoURL"];
    }
    [changeRequest commitChangesWithCompletion:^(NSError *error) {
        [self respondWith:error callbackId:command.callbackId];
    }];
}

- (void)useEmulator:(CDVInvokedUrlCommand*)command {
    NSString* host = [command.arguments objectAtIndex:0];
    id port = [command.arguments objectAtIndex:1];

    [[FIRAuth auth] useEmulatorWithHost:host port:[port intValue]];

    CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void) respondWith:(NSError*)error callbackId:(NSString*)callbackId {
    CDVPluginResult *pluginResult;
    if (error) {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.localizedDescription];
    } else {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    }
    [self.commandDelegate sendPluginResult:pluginResult callbackId:callbackId];
}

- (CDVPluginResult*)getProfileResult:(FIRUser *)user {
    NSDictionary* result = nil;
    if (user) {
        result = @{
            @"uid": user.uid,
            @"providerId": user.providerID,
            @"displayName": user.displayName ? user.displayName : @"",
            @"email": user.email ? user.email : @"",
            @"phoneNumber": user.phoneNumber ? user.phoneNumber : @"",
            @"photoURL": user.photoURL ? user.photoURL.absoluteString : @"",
            @"emailVerified": [NSNumber numberWithBool:user.emailVerified]
        };
    }

    return [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:result];
}

@end

File: src/android/FirebaseAuthenticationPlugin.java
package by.chemerisuk.cordova.firebase;

import static com.google.android.gms.tasks.Tasks.await;

import android.net.Uri;
import android.util.Log;

import by.chemerisuk.cordova.support.CordovaMethod;
import by.chemerisuk.cordova.support.ReflectiveCordovaPlugin;

import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.FacebookAuthProvider;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.FirebaseAuth.AuthStateListener;
import com.google.firebase.auth.GetTokenResult;
import com.google.firebase.auth.GoogleAuthProvider;
import com.google.firebase.auth.PhoneAuthCredential;
import com.google.firebase.auth.PhoneAuthOptions;
import com.google.firebase.auth.PhoneAuthProvider;
import com.google.firebase.auth.TwitterAuthProvider;
import com.google.firebase.FirebaseException;
import com.google.firebase.auth.UserProfileChangeRequest;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaArgs;
import org.apache.cordova.PluginResult;

import org.json.JSONException;
import org.json.JSONObject;

import static java.util.concurrent.TimeUnit.MILLISECONDS;

import androidx.annotation.NonNull;


public class FirebaseAuthenticationPlugin extends ReflectiveCordovaPlugin implements AuthStateListener {
    private static final String TAG = "FirebaseAuthentication";

    private FirebaseAuth firebaseAuth;
    private CallbackContext authStateCallback;

    @Override
    protected void pluginInitialize() {
        Log.d(TAG, "Starting Firebase Authentication plugin");
        firebaseAuth = FirebaseAuth.getInstance();
    }

    @CordovaMethod
    private void setAuthStateChanged(CordovaArgs args, CallbackContext callbackContext) throws JSONException {
        boolean disable = args.getBoolean(0);
        if (authStateCallback != null) {
            authStateCallback = null;
            firebaseAuth.removeAuthStateListener(this);
        }
        if (!disable) {
            authStateCallback = callbackContext;
            firebaseAuth.addAuthStateListener(this);
        }
    }

    @CordovaMethod
    private void getCurrentUser(CallbackContext callbackContext) {
        PluginResult pluginResult = getProfileResult(firebaseAuth.getCurrentUser());
        callbackContext.sendPluginResult(pluginResult);
    }

    @CordovaMethod
    private void getIdToken(CordovaArgs args, CallbackContext callbackContext) throws Exception {
        boolean forceRefresh = args.getBoolean(0);
        FirebaseUser user = firebaseAuth.getCurrentUser();
        if (user == null) {
            callbackContext.error("User is not authorized");
        } else {
            GetTokenResult result = await(user.getIdToken(forceRefresh));
            callbackContext.success(result.getToken());
        }
    }

    @CordovaMethod
    private void createUserWithEmailAndPassword(CordovaArgs args, CallbackContext callbackContext) throws Exception {
        String email = args.getString(0);
        String password = args.getString(1);
        await(firebaseAuth.createUserWithEmailAndPassword(email, password));
        callbackContext.success();
    }

    @CordovaMethod
    private void sendEmailVerification(CallbackContext callbackContext) throws Exception {
        FirebaseUser user = firebaseAuth.getCurrentUser();
        if (user == null) {
            callbackContext.error("User is not authorized");
        } else {
            await(user.sendEmailVerification());
            callbackContext.success();
        }
    }

    @CordovaMethod
    private void sendPasswordResetEmail(CordovaArgs args, CallbackContext callbackContext) throws Exception {
        String email = args.getString(0);
        await(firebaseAuth.sendPasswordResetEmail(email));
        callbackContext.success();
    }

    @CordovaMethod
    private void signInAnonymously(CallbackContext callbackContext) throws Exception {
        await(firebaseAuth.signInAnonymously());
        callbackContext.success();
    }

    @CordovaMethod
    private void signInWithEmailAndPassword(CordovaArgs args, CallbackContext callbackContext) throws Exception {
        String email = args.getString(0);
        String password = args.getString(1);
        await(firebaseAuth.signInWithEmailAndPassword(email, password));
        callbackContext.success();
    }

    @CordovaMethod
    private void signInWithGoogle(CordovaArgs args, CallbackContext callbackContext) throws Exception {
        String idToken = args.getString(0);
        String accessToken = args.getString(1);
        signInWithCredential(GoogleAuthProvider.getCredential(idToken, accessToken), callbackContext);
    }

    @CordovaMethod
    private void signInWithFacebook(CordovaArgs args, CallbackContext callbackContext) throws Exception {
        String accessToken = args.getString(0);
        signInWithCredential(FacebookAuthProvider.getCredential(accessToken), callbackContext);
    }

    @CordovaMethod
    private void signInWithTwitter(CordovaArgs args, CallbackContext callbackContext) throws Exception {
        String token = args.getString(0);
        String secret = args.getString(1);
        signInWithCredential(TwitterAuthProvider.getCredential(token, secret), callbackContext);
    }

    @CordovaMethod
    private void signInWithVerificationId(CordovaArgs args, CallbackContext callbackContext) throws Exception {
        String verificationId = args.getString(0);
        String code = args.getString(1);
        await(firebaseAuth.signInWithCredential(PhoneAuthProvider.getCredential(verificationId, code)));
        callbackContext.success();
    }

    @CordovaMethod
    private void verifyPhoneNumber(CordovaArgs args, CallbackContext callbackContext) throws JSONException {
        String phoneNumber = args.getString(0);
        long timeoutMillis = args.optLong(1);
        PhoneAuthOptions.Builder options = PhoneAuthOptions.newBuilder(firebaseAuth)
                .setActivity(cordova.getActivity())
                .setPhoneNumber(phoneNumber)
                .setCallbacks(new PhoneAuthProvider.OnVerificationStateChangedCallbacks() {
                    @Override
                    public void onVerificationCompleted(@NonNull PhoneAuthCredential credential) {
                        FirebaseUser user = firebaseAuth.getCurrentUser();
                        if (user != null) {
                            user.updatePhoneNumber(credential);
                        }
                    }

                    @Override
                    public void onCodeSent(@NonNull String verificationId, @NonNull PhoneAuthProvider.ForceResendingToken forceResendingToken) {
                        callbackContext.success(verificationId);
                    }

                    @Override
                    public void onVerificationFailed(@NonNull FirebaseException e) {
                        callbackContext.error(e.getMessage());
                    }
                });

        if (timeoutMillis > 0) {
            options.setTimeout(timeoutMillis, MILLISECONDS);
        }
        PhoneAuthProvider.verifyPhoneNumber(options.build());
    }

    @CordovaMethod
    private void signInWithCustomToken(CordovaArgs args, CallbackContext callbackContext) throws Exception {
        String idToken = args.getString(0);
        await(firebaseAuth.signInWithCustomToken(idToken));
        callbackContext.success();
    }

    @CordovaMethod
    private void signOut(CallbackContext callbackContext) {
        firebaseAuth.signOut();
        callbackContext.success();
    }

    @CordovaMethod
    private void setLanguageCode(CordovaArgs args, CallbackContext callbackContext) throws JSONException {
        String languageCode = args.getString(0);
        if (languageCode == null) {
            firebaseAuth.useAppLanguage();
        } else {
            firebaseAuth.setLanguageCode(languageCode);
        }
        callbackContext.success();
    }

    @CordovaMethod
    private void updateProfile(CordovaArgs args, CallbackContext callbackContext) throws Exception {
        FirebaseUser user = firebaseAuth.getCurrentUser();
        if (user == null) {
            callbackContext.error("User is not authorized");
        } else {
            JSONObject params = args.getJSONObject(0);
            UserProfileChangeRequest request = createProfileChangeRequest(params);
            await(user.updateProfile(request));
            callbackContext.success();
        }
    }

    @CordovaMethod
    private void useEmulator(CordovaArgs args, CallbackContext callbackContext) throws JSONException {
        String host = args.getString(0);
        int port = args.getInt(1);
        firebaseAuth.useEmulator(host, port);
        callbackContext.success();
    }

    private void signInWithCredential(AuthCredential credential, CallbackContext callbackContext) throws Exception {
        await(firebaseAuth.signInWithCredential(credential));
        callbackContext.success();
    }

    @Override
    public void onAuthStateChanged(@NonNull FirebaseAuth auth) {
        if (authStateCallback != null) {
            PluginResult pluginResult = getProfileResult(firebaseAuth.getCurrentUser());
            pluginResult.setKeepCallback(true);
            authStateCallback.sendPluginResult(pluginResult);
        }
    }

    private static PluginResult getProfileResult(FirebaseUser user) {
        if (user == null) {
            return new PluginResult(PluginResult.Status.OK, (String)null);
        }

        JSONObject result = new JSONObject();

        try {
            result.put("uid", user.getUid());
            result.put("displayName", user.getDisplayName());
            result.put("email", user.getEmail());
            result.put("phoneNumber", user.getPhoneNumber());
            result.put("photoURL", user.getPhotoUrl());
            result.put("providerId", user.getProviderId());
            result.put("emailVerified", user.isEmailVerified());

            return new PluginResult(PluginResult.Status.OK, result);
        } catch (JSONException e) {
            Log.e(TAG, "Fail to process getProfileData", e);

            return new PluginResult(PluginResult.Status.ERROR, e.getMessage());
        }
    }

    private static UserProfileChangeRequest createProfileChangeRequest(JSONObject jsonObject) throws JSONException {
        UserProfileChangeRequest.Builder requestBuilder = new UserProfileChangeRequest.Builder();
        if (jsonObject.has("displayName")) {
            String displayName = jsonObject.getString("displayName");
            requestBuilder = requestBuilder.setDisplayName(displayName);
        }
        if (jsonObject.has("photoURL")) {
            String photoURL = jsonObject.getString("photoURL");
            requestBuilder = requestBuilder.setPhotoUri(Uri.parse(photoURL));
        }
        return requestBuilder.build();
    }
}
